http-clients/src/main/java/org/triplea/http/client/web/socket/GenericWebSocketClient.java

```diff
@@ -19,74 +24,102 @@
-    client
-        .connect(this, errorHandler)
-        .exceptionally(
-            throwable -> {
-              log.log(
-                  Level.SEVERE, "Unexpected exception completing websocket connection", throwable);
-              return false;
-            });
+  @Builder
+  public GenericWebSocketClient(
+      @Nonnull final URI websocketUri, @Nonnull final Consumer<String> errorHandler) {
+    this.websocketUri = websocketUri;
+    this.errorHandler = errorHandler;
   }
 
   /**
    * Non-blocking send of a message to the server. Implementation note: data is sent as a JSON
    * string, this method handles conversion of the parameter object to JSON.
    *
    * @param message The data object to send to the server.
+   * @deprecated Call {@code sendMessage(WebsocketMessage)} instead.
    */
+  @Deprecated
   public void send(final ClientMessageEnvelope message) {
     // we get by doing the send on a new thread.
-    client.sendMessage(gson.toJson(message));
+    webSocketConnection.sendMessage(gson.toJson(message));
   }
 
-  /**
-   * Removes connection lost listeners and starts a non-blocking close of the websocket connection.
-   */
+  @Override
+  public void connect() {
+    addListener(ServerErrorMessage.TYPE, message -> errorHandler.accept(message.getError()));
+    webSocketConnection = new WebSocketConnection(swapHttpsToWssProtocol(websocketUri));
+    webSocketConnection.connect(this, errorHandler);
+  }
+
+  @VisibleForTesting
+  static URI swapHttpsToWssProtocol(final URI uri) {
+    return uri.getScheme().equals("https")
+        ? URI.create(uri.toString().replace("https", "wss"))
+        : uri;
+  }
+
+  /** Starts a non-blocking close of the websocket connection. */
+  @Override
   public void close() {
-    client.close();
+    webSocketConnection.close();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends WebSocketMessage> void addListener(
+      final MessageType<T> messageType, final Consumer<T> messageHandler) {
+
+    final Consumer<Object> messageConsumer = object -> messageHandler.accept((T) object);
+    listeners.add(
+        MessageListener.<T>builder() //
+            .messageType(messageType)
+            .listener(messageConsumer)
+            .build());
+  }
+
+  @Override
+  public void sendMessage(final WebSocketMessage message) {
+    webSocketConnection.sendMessage(gson.toJson(message.toEnvelope()));
   }
 
+  @Override
   public void addConnectionClosedListener(final Runnable connectionClosedListener) {
     connectionClosedListeners.add(connectionClosedListener);
   }
 
   @Override
   public void messageReceived(final String message) {
-    final ServerMessageEnvelope converted = gson.fromJson(message, ServerMessageEnvelope.class);
-    messageListener.accept(converted);
+    final MessageEnvelope converted = gson.fromJson(message, MessageEnvelope.class);
+
+    listeners.stream()
+        .filter(listener -> converted.messageTypeIs(listener.messageType))
```

Comment:
Thoughts about using a MultiMap to get the `O(1)` runtime of this frequently used method?

Reply:
I did want a map, the listeners are of mixed type. If you can see how to solve that, it would make for cleaner logic probably. Otherwise, the mixed typing caused issues defining such a map.

FWIW, I don't think the iteration and lookup is necessarily a  problem. A recent saying I've come across that I found interesting, https://blog.cloudflare.com/when-bloom-filters-dont-bloom/, 
> "One colleague often says: "You can assume modern CPUs are infinitely fast. They run at infinite speed until they hit the memory wall"."

In this case, we are not going to have all listeners loaded from main memory, but still the point remains that a map is not necessarily going to be that much faster than sequential scanning on a pretty limited collection size.

Reply:
I agree that a map might not yield the desired performance benefits with this few listeners, but your answer about the types surprised me:
What would speak against code that looks something along the lines of this:
```java
private final Multimap<MessageType<?>, MessageListener> listeners = HashMultimap.create();

@Override
  public void messageReceived(final String message) {
    final MessageEnvelope converted = gson.fromJson(message, MessageEnvelope.class);

    listeners.get(converted.getMessageType()).forEach(listener ->
                listener.listener.accept(
                    converted.getPayload(listener.messageType.getPayloadType())));
```

Reply:
Seems you'll lose type coherence, `getMessageType()` returns a string. `MessageListener` is also a generic type as well, it should match the generic type for `MessageType`
 
One also needs to account for the adding a listener as well under those constraints. The type safety we worked pretty hard for previously, clarity and correctness are important heuristics. We also need to think about system level issues, whether the incoming/outgoing message pattern is going to be clear and sustainable.